sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/specialized/BlockBlobApiTests.java

```diff
@@ -0,0 +1,2551 @@
+            parallelTransferOptions, true);
+
+        StepVerifier.create(uploadOperation.then(collectBytesInBuffer(blockBlobAsyncClient.download())))
+            .assertNext(it -> assertTrue(compareListToBuffer(dataList, it)))
+            .verifyComplete();
+
+        StepVerifier.create(blockBlobAsyncClient.listBlocks(BlockListType.ALL))
+            .assertNext(it -> assertEquals(it.getCommittedBlocks().size(), blockCount))
+            .verifyComplete();
+    }
+
+    private static Stream<Arguments> bufferedUploadChunkedSourceSupplier() {
+        return Stream.of(
+            // First item fits entirely in the buffer, next item spans two buffers
+            Arguments.of(new int[]{7, 7}, 10, 2, 2),
+            // Multiple items fit non-exactly in one buffer.
+            Arguments.of(new int[]{3, 3, 3, 3, 3, 3, 3}, 10, 2, 3),
+            // Data fits exactly and does not need chunking.
+            Arguments.of(new int[]{10, 10}, 10, 2, 2),
+            // Data needs chunking and does not fit neatly in buffers. Requires waiting for buffers to be released.
+            Arguments.of(new int[]{50, 51, 49}, 10, 2, 15));
+    }
+
+    @ParameterizedTest
+    @MethodSource("bufferedUploadHandlePathingSupplier")
+    @EnabledIf("com.azure.storage.blob.BlobTestBase#isLiveMode")
+    public void bufferedUploadHandlePathing(int[] dataSizeList, int blockCount) {
+        List<ByteBuffer> dataList = new ArrayList<>();
+        for (int size : dataSizeList) {
+            dataList.add(getRandomData(size * Constants.MB));
+        }
+        Mono<BlockBlobItem> uploadOperation = blobAsyncClient.upload(Flux.fromIterable(dataList),
+            new ParallelTransferOptions().setMaxSingleUploadSizeLong(4L * Constants.MB), true);
+
+        StepVerifier.create(uploadOperation.then(collectBytesInBuffer(blockBlobAsyncClient.download())))
+            .assertNext(it -> assertTrue(compareListToBuffer(dataList, it)))
+            .verifyComplete();
+
+        StepVerifier.create(blockBlobAsyncClient.listBlocks(BlockListType.ALL))
+            .assertNext(it -> assertEquals(it.getCommittedBlocks().size(), blockCount))
+            .verifyComplete();
+    }
+
+    private static Stream<Arguments> bufferedUploadHandlePathingSupplier() {
+        return Stream.of(
+            Arguments.of(new int[]{4 * Constants.MB + 1, 10}, 2),
+            Arguments.of(new int[]{4 * Constants.MB}, 0),
+            Arguments.of(new int[]{10, 100, 1000, 10000}, 0),
+            Arguments.of(new int[]{4 * Constants.MB, 4 * Constants.MB}, 2));
+    }
+
+    @ParameterizedTest
+    @MethodSource("bufferedUploadHandlePathingSupplier")
+    @EnabledIf("com.azure.storage.blob.BlobTestBase#isLiveMode")
+    public void bufferedUploadHandlePathingHotFlux(int[] dataSizeList, int blockCount) {
+        List<ByteBuffer> dataList = new ArrayList<>();
+        for (int size : dataSizeList) {
+            dataList.add(getRandomData(size * Constants.MB));
```

Comment:
This is going to create massively large `ByteBuffer`s on the last iteration of the test run as `dataSizeList` is `[ 4MB, 4MB ]` which means there will be buffers `[ 4TB, 4TB ]`